src/styles/text/text.js

```diff
@@ -0,0 +1,371 @@
+            let text_infos = texts[style];
+
+            for (let text in text_infos) {
+                let info = text_infos[text];
+
+                this.setFont(tile, info.text_style);
+                this.drawText(text, info.position, tile);
+
+                info.texcoords = Builders.getTexcoordsForSprite(
+                    info.position,
+                    info.size,
+                    texture_size
+                );
+            }
+        }
+    },
+
+    // Called on main thread from worker, to create atlas of labels for a tile
+    addTexts (tile, texts) {
+        this.texts[tile] = texts;
+
+        let texture_size = this.setTextureTextPositions(tile, texts);
+        let context = this.canvas[tile].context;
+
+        console.log(`text summary for tile ${tile}: fits in ${texture_size[0]}x${texture_size[1]}px`);
+
+        // update the canvas "context"
+        this.canvas[tile].canvas.width = texture_size[0];
+        this.canvas[tile].canvas.height = texture_size[1];
+        context.clearRect(0, 0, texture_size[0], texture_size[1]);
+
+        // create a texture
+        this.texture[tile] = new Texture(this.gl, 'labels-' + tile, { filtering: 'linear' });
+        this.texture[tile].setCanvas(this.canvas[tile].canvas);
+
+        // ask for rasterization for the text set
+        this.rasterize(tile, texts, texture_size);
+
+        this.texture[tile].update();
+
+        return Promise.resolve(this.texts[tile]);
+    },
+
+    // Override
+    startData () {
+        let tile_data = this.super.startData.apply(this);
+        tile_data.queue = [];
+        return tile_data;
+    },
+
+    // Override
+    endData (tile_data) {
+        // Count collected text
+        let tile, count;
+        if (tile_data.queue.length > 0) {
+            tile = tile_data.queue[0][2].tile.key;
+            count = Object.keys(this.texts[tile]||{}).length;
+            console.log(`# texts for tile ${tile}: ${count}`);
+        }
+        if (!count) {
+            return Promise.resolve();
+        }
+
+        // Attach tile-specific label atlas to mesh as a texture uniform
+        tile_data.uniforms = { u_textures: ['labels-'+tile] };
+
+        // first call to main thread, ask for text pixel sizes
+        return WorkerBroker.postMessage('Text', 'getTextSizes', tile, this.texts[tile]).then(texts => {
+            this.bboxes[tile] = [];
+
+            // cleanup of texts that should be removed after occlusion test
+            for (let style in texts) {
+                let text_infos = texts[style];
+
+                for (let text in text_infos) {
+                    let text_info = text_infos[text];
+                    let label;
+                    let keep_in_tile;
+                    let move_in_tile;
+                    let geometry = this.geometries[tile][style][text];
+
+                    if (geometry.type === "LineString") {
+                        let exceed_heuristic = this.label_style.lines.exceed;
```

Comment:
`let` variables are scoped to the block, so `exceed_heuristic` isn't defined when referenced in the `discard` call below. You can move the `let exceed_heuristic` declaration outside this block, like `keep_in_tile` etc.